app/unit-tests/src/org/commcare/android/tests/formsave/CyclicCasesPurgeTest.java

```diff
@@ -35,17 +39,27 @@ public void setup() {
     }
 
     @Test
-    public void testFormSaveResultingIntoCaseCycles_ShouldFail() {
+    public void testFormSaveResultingIntoCaseCycles_ShouldFail() throws InvalidCaseGraphException {
         TestUtils.processResourceTransactionIntoAppDb("/inputs/case_create_for_cyclic_case_purge.xml");
         AndroidSessionWrapper sessionWrapper =
                 CommCareApplication.instance().getCurrentSessionWrapper();
         CommCareSession session = sessionWrapper.getSession();
         session.setDatum("case_id", "fe9adf5c-7661-4f8e-a48e-4059fcf1f9dd");
         session.setCommand("m2-f0");
+
+        // Completes a form that create a cycle in case indices
         FormEntryActivity formEntryActivity = ActivityLaunchUtils.launchFormEntry();
         View finishButton = formEntryActivity.findViewById(R.id.nav_btn_finish);
         finishButton.performClick();
+
+        // verify that form save results into an error
         String message = ((TextView)formEntryActivity.getCurrentAlertDialog().getUnderlyingDialog().getDialog().findViewById(R.id.dialog_message)).getText().toString();
         assert message.contentEquals(formEntryActivity.getString(R.string.invalid_case_graph_error));
+
+        // Verify that the form record was quarantined
+        CommCareApplication.instance().getCurrentSessionWrapper().getFormRecord().getStatus().contentEquals(STATUS_QUARANTINED);
+
+        // Verify that the normal purge succeeds without error
+        CaseUtils.purgeCases();
```

Comment:
Nit - I'd probably try/catch this and throw a more explicit error so that if a change was made which didn't protect the db state, the failure they'd get can be contextualized more clearly (IE: The failure was "CommCare didn't rollback / recovery from a form that created an invalid state" rather than "there was a cycle")

Reply:
https://github.com/dimagi/commcare-android/pull/2504/commits/3cd7f5f0ab8b933e65ad60f6488bbd93591b07bb